File: 8.字符串转换整数-atoi.js
/*
 * @lc app=leetcode.cn id=8 lang=javascript
 *
 * [8] 字符串转换整数 (atoi)
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function(s) {
  const match = s.trim().match(/^(-|\+)?\d+/);
  if (!match) return 0;
  const num = Number(match[0]);
  const max = 2 ** 31 - 1;
  const min = -(2 ** 31);
  return Math.max(Math.min(num, max), min);
};
// @lc code=end



// myAtoi('   -42');
myAtoi('4193 with words');

File: 8.字符串转换整数-atoi.java
/*
 * @lc app=leetcode.cn id=8 lang=java
 *
 * [8] 字符串转换整数 (atoi)
 */

// @lc code=start
class Solution {
    public int myAtoi(String s) {
        char[] c = s.trim().toCharArray();

        if (c.length == 0)
            return 0;

        int res = 0, bndry = Integer.MAX_VALUE / 10;

        int i = 1, sign = 1;

        if (c[0] == '-') {
            sign = -1;
        } else if (c[0] != '+') {
            i = 0;
        }

        for (int j = i; j < c.length; j++) {
            if (c[j] < '0' || c[j] > '9') {
                break;
            }

            if (res > bndry || res == bndry && c[j] > '7') {
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }

            res = res * 10 + (c[j] - '0');
        }

        return sign * res;
    }
}
// @lc code=end

class MainSolutionmyAtoi {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.myAtoi("-91283472332"));
    }
}
